pkg/scheduler/plugins/gang/gang_plugin.go

```diff
@@ -0,0 +1,507 @@
+			waitingGangId := GetNamespaceSplicingName(waitingPod.GetPod().Namespace,
+				GetGangNameByPod(waitingPod.GetPod()))
+			if waitingGangId == gangId {
+				klog.Errorf("postFilter rejects the pod, gangName:%v, podName:%v",
+					GetNamespaceSplicingName(pod.Namespace, pod.Name), gangId)
+				waitingPod.Reject(p.Name(), "gang rejection in PostFilter")
+			}
+		})
+		gang.SetScheduleCycleValid(false)
+		return &framework.PostFilterResult{}, framework.NewStatus(framework.Unschedulable,
+			fmt.Sprintf("Gang %v gets rejected this cycle due to Pod %v is unschedulable even after "+
+				"PostFilter in StrictMode", gangId, pod.Name))
+	}
+
+	return &framework.PostFilterResult{}, framework.NewStatus(framework.Success, "")
+}
+
+// Permit
+//we will calculate all Gangs in GangGroup whether the current number of assumed-pods in each Gang meets the Gang's minimum requirement.
+//and decide whether we should let the pod wait in Permit stage or let the whole gangGroup go binding
+func (p *GangPlugin) Permit(ctx context.Context, state *framework.CycleState, pod *corev1.Pod, nodeName string) (*framework.Status, time.Duration) {
+	waitTime, s := p.PermitCheck(pod)
+	var retStatus *framework.Status
+	switch s {
+	case NotFoundInCache:
+		return framework.NewStatus(framework.Unschedulable, "Gang not found in gangCache"), 0
+	case Wait:
+		klog.Infof("Pod %v from gang %v is waiting to be scheduled at Permit stage",
+			GetNamespaceSplicingName(pod.Namespace, pod.Name),
+			GetNamespaceSplicingName(pod.Namespace,
+				GetGangNameByPod(pod)))
+		retStatus = framework.NewStatus(framework.Wait)
+		p.ActivateGang(pod, state)
+	case Success:
+		p.AllowGangGroup(pod)
+		retStatus = framework.NewStatus(framework.Success)
+		waitTime = 0
+	}
+	return retStatus, waitTime
+}
+
+// Reserve is the functions invoked by the framework at "reserve" extension point.
+func (p *GangPlugin) Reserve(ctx context.Context, state *framework.CycleState, pod *corev1.Pod, nodeName string) *framework.Status {
+	return nil
+}
+
+// Unreserve
+//(1) handle the timeout gang
+//(2) do nothing when bound failed
+func (p *GangPlugin) Unreserve(ctx context.Context, state *framework.CycleState, pod *corev1.Pod, nodeName string) {
+	gangId := GetNamespaceSplicingName(pod.Namespace,
+		GetGangNameByPod(pod))
+	gang := p.gangCache.GetGangFromCache(gangId)
+	if gang == nil {
+		klog.Errorf("can't find gang, gangName:%v, podName:%v", gangId,
+			GetNamespaceSplicingName(pod.Namespace, pod.Name))
+		return
+	}
+
+	// gang time out
+	if !gang.IsGangResourceSatisfied() {
+		klog.Infof("gang is time out,start to release the assumed resource and add annotations to the "+
+			"gang's children, gangName:%v, podName:%v", gangId, GetNamespaceSplicingName(pod.Namespace, pod.Name))
+		pods, err := p.podLister.List(labels.NewSelector())
+		if err != nil {
+			klog.Errorf("unReserve list  pod err:%v", err.Error())
+			return
+		}
+		for _, podOriginal := range pods {
```

Comment:
if pod is preempted, not timeout,  we should not update timeout annotation; how can we distinguish timeout or preemtion? may be we can record a firstPremit timestamp on gang, and when unreserve, we judge curtime -  firstPremit  >= waitTime, is yes, no matter is preempted or timeout, we should update timeout annotation.